tensorflow/core/kernels/stack.cc

```diff
@@ -72,19 +87,86 @@ class Stack : public ResourceBase {
       return errors::InvalidArgument("Stack[", stack_name_,
                                      "] is empty when calling Pop().");
     }
-    *value = stack_.back();
+    size_t index = stack_.size() - 1;
+    while (is_swapping_ins_[index]) {
+      cond_swapping_ins_[index].wait(l);
+      index = stack_.size() - 1;
+    }
+    *value = stack_.back().value;
+    if (stack_.back().unswap_iter) {
+      unswapped_lru_.erase(*(stack_.back().unswap_iter));
+    }
+    if (stack_.back().swap_iter) {
+      swapped_mru_.erase(*(stack_.back().swap_iter));
+    }
     stack_.pop_back();
     return Status::OK();
   }
 
+  Status MarkAsSwapped(TensorAndAllocation* value, Tensor* cpu_tensor) {
+    mutex_lock l(mu_);
+    TF_RETURN_IF_ERROR(CheckNotClosed());
+
+    value->tensor = *cpu_tensor;
+    value->swapped_to_cpu = true;
+
+    return Status::OK();
+  }
+
+  Status MarkAsUnswapped(TensorAndAllocation* value, Tensor* device_tensor) {
+    mutex_lock l(mu_);
+    TF_RETURN_IF_ERROR(CheckNotClosed());
+
+    value->tensor = *device_tensor;
+    value->swapped_to_cpu = false;
+
+    return Status::OK();
+  }
+
+  bool GetTensorToSwapOut(std::function<bool(const TensorAndAllocation&)> cond,
+                          TensorAndAllocation** value) {
+    mutex_lock l(mu_);
+    while (!unswapped_lru_.empty()) {
+      size_t index = unswapped_lru_.front();
+      unswapped_lru_.pop_front();
+      stack_[index].unswap_iter.reset();
+      if (!cond(stack_[index].value)) {
+        continue;
+      }
+      swapped_mru_.push_back(index);
+      stack_[index].swap_iter = std::prev(swapped_mru_.end());
+      *value = &(stack_[index].value);
+      return true;
+    }
+    return false;  // nothing to swap out
+  }
+
+  bool GetTensorToSwapIn(TensorAndAllocation** value, size_t* index) {
+    mutex_lock l(mu_);
+    while (!swapped_mru_.empty()) {
+      *index = swapped_mru_.back();
+      swapped_mru_.pop_back();
+      stack_[*index].swap_iter.reset();
+      *value = &(stack_[*index].value);
+      is_swapping_ins_[*index] = true;
+      return true;
+    }
+    return false;  // nothing to swap in
+  }
+
+  void FinishSwappingIn(size_t index) {
+    mutex_lock l(mu_);
+    is_swapping_ins_[index] = false;
+    cond_swapping_ins_[index].notify_one();
+  }
+
   // We don't swap the first tensor on the stack and any subsequent tensors
   // that share the buffer with the first tensor.
   bool IsUsefulToSwap(const Tensor& tensor) const {
-    mutex_lock l(mu_);
     if (stack_.empty()) {
```

Comment:
this line is failing internal test , here is the error : 
` error: reading variable 'stack_' requires holding mutex 'mu_' [-Werror,-Wthread-safety-analysis]`

Reply:
Actually, this was intentional, since `IsUsefulToSwap` is now only called within `GetTensorToSwapOut` (as a part of its anonymous function parameter), which already locks a mutex, locking mutex again here would cause deadlock. But I'll find a way around this, thank you for posting the errors.

Reply:
After thinking about this for a while, I couldn't find a good solution to keep `IsUsefulToSwap` without recursive mutex. So, I've opted to inline the function's body inside `GetTensorToSwapOut` instead. I'm open to suggestions, thank you.